network/p2p/cache/node_blocklist_wrapper.go

```diff
@@ -0,0 +1,191 @@
+package cache
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/id"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/operation"
+)
+
+// IdentifierSet represents a set of node IDs (operator-defined) whose communication should be blocked.
+type IdentifierSet map[flow.Identifier]struct{}
+
+// Includes returns true iff id ∈ s
+func (s IdentifierSet) Contains(id flow.Identifier) bool {
+	_, found := s[id]
+	return found
+}
+
+// NodeBlocklistWrapper is a wrapper for an `id.IdentityProvider` instance, where the
+// wrapper overrides the `Ejected` flag to true for all NodeIDs in a `blocklist`.
+// To avoid modifying the source of the identities, the wrapper creates shallow copies
+// of the identities (whenever necessary) and modifies the `Ejected` flag only in
+// the copy.
+// The `NodeBlocklistWrapper` internally represents the `blocklist` as a map, to enable
+// performant lookup. However, the exported API works with `flow.IdentifierList` for
+// blocklist, as this is a broadly supported data structure which lends itself better
+// to config or command-line inputs.
+type NodeBlocklistWrapper struct {
+	m  sync.RWMutex
+	db *badger.DB
+
+	identityProvider id.IdentityProvider
+	blocklist        IdentifierSet // `IdentifierSet` is a map, hence efficient O(1) lookup
+}
+
+var _ id.IdentityProvider = (*NodeBlocklistWrapper)(nil)
+
+// NewNodeBlocklistWrapper wraps the given `IdentityProvider`. The blocklist is
+// loaded from the database (or assumed to be empty if no database entry is present).
+func NewNodeBlocklistWrapper(identityProvider id.IdentityProvider, db *badger.DB) (*NodeBlocklistWrapper, error) {
+	blocklist, err := retrieveBlocklist(db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read set of blocked node IDs from data base: %w", err)
+	}
+
+	return &NodeBlocklistWrapper{
+		db:               db,
+		identityProvider: identityProvider,
+		blocklist:        blocklist,
+	}, nil
+}
+
+// Update sets the wrapper's internal set of blocked nodes to `blocklist`. Empty list and `nil`
+// (equivalent to empty list) are accepted inputs. To avoid legacy entries in the data base, this
+// function purges the entire data base entry if `blocklist` is empty.
+// This implementation is _eventually consistent_, where changes are written to the data base first
+// and then (non-atomically!) the in-memory set of blocked nodes is updated. This strongly
+// benefits performance and modularity. No errors are expected during normal operations.
+func (w *NodeBlocklistWrapper) Update(blocklist flow.IdentifierList) error {
+	b := blocklist.Lookup() // convert slice to
+	err := persistBlocklist(b, w.db)
+	if err != nil {
+		return fmt.Errorf("failed to persist set of blocked nodes to the data base: %w", err)
+	}
+
+	w.m.Lock()
+	w.blocklist = b
+	w.m.Unlock()
```

Comment:
Why not lock before updating database? 

I'm not sure if this is concurrent safe. Consider the following events:

1. process A calls `Update(N1, N2)`
2. process B concurrently calls `Update(N3, N4)`
3. process A persist database with [N1, N2]
4. process B persist database with [N3, N4]
5. process B updates w.blocklist with [N3, N4]
6. process A updates w.blocklist with [N1, N2]

Now database has `[N3, N4]`, but w.blocklist has `[N1, N2]`, they are inconsistent.

Reply:
Good point. I didn't pay too much attention to this aspect because we are intending to update the wrapper state manually (through an admin command). 
But you are right, better safe than inconsistent. Since calls to `Update` are going to be _very_ infrequent there are no practical performance concerns with doing the data base write within the lock.

Reply:
revision: https://github.com/onflow/flow-go/pull/3344/commits/bd14684c0656b18aa4e22365d200bb94a7229fb7